librz/io/p/io_srec.c

```diff
@@ -0,0 +1,810 @@
+				
+				eol = strchr(str + 1, 'S');
+				if (eol) {
+					*eol = 0;
+				}
+
+				cksum = bc;
+				cksum += addr_tmp >> 8;
+				cksum += addr_tmp >> 16;
+				cksum += addr_tmp >> 32;
+				cksum += addr_tmp & 0xff;
+
+				if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+					//previous block is not contiguous, or
+					//section buffer is full => write a sparse chunk
+					if (sec_size && sec_size < UT16_MAX) {
+						if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (int)sec_size) != sec_size) {
+							eprintf("sparse buffer problem, giving up\n");
+							goto fail;
+						}
+					}
+					//advance cursor, reset section
+					sec_start = segreg + addr_tmp;
+					next_addr = addr_tmp;
+					sec_size = 0;
+				}
+
+				for (i = 0; i < bc - 5; i++) {
+					if (sscanf(str + 12 + (i * 2), "%02x", &byte) != 1) {
+						eprintf("unparsable data !\n");
+						goto fail;
+					}
+					if (sec_size + i < sec_count) {
+						sec_tmp[sec_size + i] = (ut8)byte & 0xff;
+					}
+					cksum += byte;
+				}
+				sec_size += bc;
+				next_addr += bc;
+				if (eol) {
+					// checksum
+					if (sscanf(str + 12 + (i * 2), "%02x", &byte) != 1) {
+						eprintf("unparsable data !\n");
+						goto fail;
+					}
+					cksum = ~(cksum & 0xff); //this might be useless cause cksum is ut8
+					if (cksum != byte) {
+						eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+						goto fail;
+					}
+					*eol = 'S';
+				}
+				str = eol;
+				break;
+
+			case 7: //same with S9 but on 32 bits
+				
+				// S / 7 / (addres+data+checksum) bytes / address / - / checksum
+				l = sscanf(str + 4, "%08x", &addr_tmp);			
+				addr_tmp &= 0xffffffff;
+				
+				eol = strchr(str + 1, 'S');
+				if (eol) {
+					*eol = 0;
+				}
+
+				cksum = bc;
+				cksum += addr_tmp >> 8;
+				cksum += addr_tmp >> 16;
+				cksum += addr_tmp >> 32;
+				cksum += addr_tmp & 0xff;
+
+				if ((next_addr != addr_tmp) || ((sec_size + bc) > SEC_MAX)) {
+					//previous block is not contiguous, or
+					//section buffer is full => write a sparse chunk
+					if (sec_size && sec_size < UT16_MAX) {
+						if (rz_buf_write_at(rbuf, sec_start, sec_tmp, (int)sec_size) != sec_size) {
+							eprintf("sparse buffer problem, giving up\n");
+							goto fail;
+						}
+					}
+					//advance cursor, reset section
+					sec_start = segreg + addr_tmp;
+					next_addr = addr_tmp;
+					sec_size = 0;
+				}
+				sec_size += bc;
+				next_addr += bc;
+				if (eol) {
+					// checksum
+					if (sscanf(str + 12 + (i * 2), "%02x", &byte) != 1) {
+						eprintf("unparsable data !\n");
+						goto fail;
+					}
+					cksum = ~(cksum & 0xff); //this might be useless cause cksum is ut8
+					if (cksum != byte) {
+						eprintf("Checksum failed (got %02x expected %02x)\n", byte, cksum);
+						goto fail;
+					}
+					*eol = 'S';
+				}
+				str = eol;
+				break;
+
+			case 4: //reserved
+			}
+	} while (str);
+	free(sec_tmp);
+	return true;
+fail:
+	free(sec_tmp);
+	return false;
```

Comment:
you can simplify this by having a `bool ret = false;` and then here this
```
	} while (str);
	ret = true;
fail:
	free(sec_tmp);
	return ret;
```

Reply:
I don't understand what you mean by this. But I changed the other things

Reply:
it's a code optimization. ignore it.